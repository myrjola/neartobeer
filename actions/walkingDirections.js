import polyline from 'google-polyline';

import checkStatus from '../util/checkStatus';

export const REQUEST_WALKING_DIRECTIONS = 'REQUEST_WALKING_DIRECTIONS';
export const RECEIVE_WALKING_DIRECTIONS = 'RECEIVE_WALKING_DIRECTIONS';
export const ERROR_WALKING_DIRECTIONS = 'ERROR_WALKING_DIRECTIONS';

export const requestWalkingDirections = destination => ({
  type: REQUEST_WALKING_DIRECTIONS,
  destination,
});

export const receiveWalkingDirections = (origin, coords, destination) => ({
  type: RECEIVE_WALKING_DIRECTIONS,
  origin,
  coords,
  destination,
});

function decodeGoogleDirectionsPolyline(polylineString) {
  const coordinates = polyline.decode(polylineString);
  return coordinates.map(coordinate => ({ latitude: coordinate[0], longitude: coordinate[1] }));
}

export const errorWalkingDirections = error => ({
  type: ERROR_WALKING_DIRECTIONS,
  error,
});

export function coordinatesMatch(coord1, coord2) {
  return coord1.latitude === coord2.latitude && coord1.longitude === coord2.longitude;
}

function coordinatesToString(coord) {
  return `${coord.latitude},${coord.longitude}`;
}

function fetchWalkingDirections(destination) {
  return (dispatch) => {
    dispatch(requestWalkingDirections(destination));
    navigator.geolocation.getCurrentPosition(
      (position) => {
        const origin = position.coords;
        // FIXME: Figure out how not to repeat ourselves. This key is also in AndroidManifests.xml.
        const APIKEY = '***REMOVED***';
        const url = `https://maps.googleapis.com/maps/api/directions/json?origin=${coordinatesToString(origin)}` +
              `&destination=${coordinatesToString(destination)}&key=${APIKEY}&mode=walking`;
        return fetch(url)
          .then(checkStatus)
          .then(response => response.json())
          .then(json => decodeGoogleDirectionsPolyline(json.routes[0].overview_polyline.points))
          .then(coords => dispatch(receiveWalkingDirections(origin, coords, destination)))
          .catch(error => dispatch(errorWalkingDirections(error)));
      },
      error => dispatch(errorWalkingDirections(error)),
    );
  };
}

function shouldFetchWalkingDirections(state, destination) {
  const destinationHasChanged = coordinatesMatch(destination, state.walkingDirections.destination);
  if (!state.walkingDirections) {
    return true;
  } else if (!destinationHasChanged && state.walkingDirections.isFetching) {
    return false;
  }
  return true;
}

export function fetchWalkingDirectionsIfNeeded(destination) {
  return (dispatch, getState) => {
    if (shouldFetchWalkingDirections(getState(), destination)) {
      return dispatch(fetchWalkingDirections(destination));
    }
  };
}


// Local Variables:
// mode: js2-jsx
// End:
