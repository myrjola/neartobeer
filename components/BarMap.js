import React, { PropTypes } from 'react';
import { StyleSheet } from 'react-native';
import MapView from 'react-native-maps';

import BarMarkerContainer from '../containers/BarMarkerContainer';
import WalkingDirections from '../containers/WalkingDirections';

const mapStyle = StyleSheet.create({ flex: 1 });

const BarMap = ({ items }) => (
  <MapView
    ref={(ref) => { this.map = ref; }}
    style={mapStyle}
    showsUserLocation={true}
    showsMyLocationButton={false}
    initialRegion={{
      latitude: 59.329323,
      longitude: 18.068580,
      latitudeDelta: 0.0922,
      longitudeDelta: 0.0421,
    }}
  >
    {
      items.map(
        bar => (
          <BarMarkerContainer
            key={bar.post_id}
            bar={bar}
          />
        ),
      )
    }
    <WalkingDirections />
  </MapView>
);

BarMap.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    post_id: PropTypes.number.isRequired,
  }).isRequired),
  isFetching: PropTypes.bool.isRequired,
};

BarMap.defaultProps = {
  items: [],
};

export default BarMap;

// Local Variables:
// mode: js2-jsx
// End:
