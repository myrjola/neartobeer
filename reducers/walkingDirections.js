import { REQUEST_WALKING_DIRECTIONS, RECEIVE_WALKING_DIRECTIONS, ERROR_WALKING_DIRECTIONS,
         coordinatesMatch } from '../actions/walkingDirections';

export const initialState = {
  isFetching: false,
  error: '',
  origin: {
    longitude: 0,
    latitude: 0,
  },
  destination: {
    longitude: 0,
    latitude: 0,
  },
  coords: [],
};

function walkingDirections(state = initialState, action) {
  switch (action.type) {
    case REQUEST_WALKING_DIRECTIONS:
      return Object.assign({}, state, {
        isFetching: true,
        destination: action.destination,
      });
    case RECEIVE_WALKING_DIRECTIONS: {
      const responseCorrespondsToLatestRequest = coordinatesMatch(state.destination, action.destination);
      if (responseCorrespondsToLatestRequest) {
        return Object.assign({}, state, {
          isFetching: false,
          origin: action.origin,
          coords: action.coords,
        });
      }
      return state;
    }
    case ERROR_WALKING_DIRECTIONS:
      return Object.assign({}, state, {
        isFetching: false,
        coords: [],
        lastUpdated: action.receivedAt,
        error: action.error,
      });
    default:
      return state;
  }
}

export default walkingDirections;

// Local Variables:
// mode: js2-jsx
// End:
