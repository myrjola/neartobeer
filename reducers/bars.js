import { REQUEST_BARS, RECEIVE_BARS, INVALIDATE_BARS } from '../actions';

const BARS = [
  {
    post_id: 38,
    post_title: 'MacLarens',
    post_author: 1,
    post_content: 'Serves chilled Norrlands Guld on tap for only 29:-.' +
      'One of the most best places in town for cheap beer lovers.',
    post_category: 'Beer under 30',
    default_category: 'Beer under 30',
    post_tags: '',
    post_type: 'gd_place',
    post_status: 'publish',
    is_featured: '',
    geodir_video: '',
    post_address: 'Norrtullsgatan 11',
    post_city: 'Stockholm',
    post_region: 'Stockholm County',
    post_country: 'Sweden',
    post_zip: '113 29',
    post_latitude: 59.3419634,
    post_longitude: 18.051719899999966,
    geodir_timing: '',
    geodir_contact: '',
    geodir_email: '',
    geodir_website: '',
    geodir_twitter: '',
    geodir_facebook: '',
    geodir_special_offers: '',
    IMAGE: '',
  },
  {
    post_id: 39,
    post_title: 'Lion Bar Sveavägen 39',
    post_author: 1,
    post_content: 'The famous Lion Bar chain is a perfect find for bargain brew lovers! 7 Places all around' +
      ' Stockholm serve chilled Spendrups (40cl, tapped) for 23:- until 20h. After this the prices will increase ' +
      'in two steps: until midnight the price will be 36:-, thereafter 40:-. Go and get there early to get yourself' +
      ' a crispy brew for under 30:- !',
    post_category: 'Beer under 50',
    default_category: 'Beer under 30',
    post_tags: '',
    post_type: 'gd_place',
    post_status: 'publish',
    is_featured: '',
    geodir_video: '',
    post_address: 'Sveavägen 39 Stockholm',
    post_city: 'Stockholm',
    post_region: 'Stockholm County',
    post_country: 'Sweden',
    post_zip: '',
    post_latitude: 59.3404626,
    post_longitude: 18.05888660000005,
    geodir_timing: '',
    geodir_contact: '',
    geodir_email: '',
    geodir_website: '',
    geodir_twitter: '',
    geodir_facebook: '',
    geodir_special_offers: '',
    IMAGE: '',
  },
];

function bars(state = {
  isFetching: false,
  didInvalidate: false,
  items: BARS,
}, action) {
  switch (action.type) {
    case INVALIDATE_BARS:
      return Object.assign({}, state, {
        didInvalidate: true,
      });
    case REQUEST_BARS:
      return Object.assign({}, state, {
        isFetching: true,
        didInvalidate: false,
      });
    case RECEIVE_BARS:
      return Object.assign({}, state, {
        isFetching: false,
        didInvalidate: false,
        items: action.bars,
        lastUpdated: action.receivedAt,
      });
    default:
      return state;
  }
}

export default bars;

// Local Variables:
// mode: js2-jsx
// End:
