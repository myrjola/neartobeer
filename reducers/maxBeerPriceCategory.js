import { SET_BEER_MAX_PRICE_CATEGORY } from '../actions';

export const BEER_UNDER_30 = 'Beer under 30';
export const BEER_UNDER_40 = 'Beer under 40';
export const BEER_UNDER_50 = 'Beer under 50';
export const maxBeerPriceCategories = [BEER_UNDER_30, BEER_UNDER_40, BEER_UNDER_50];

const maxBeerPriceCategory = (state = BEER_UNDER_50, action) => {
  switch (action.type) {
    case SET_BEER_MAX_PRICE_CATEGORY:
      return action.category;
    default:
      return state;
  }
};

export default maxBeerPriceCategory;

// Local Variables:
// mode: js2-jsx
// End:
