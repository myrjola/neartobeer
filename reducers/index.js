import { combineReducers } from 'redux';
import maxBeerPriceCategory from './maxBeerPriceCategory';
import bars from './bars';

const beersUnderXApp = combineReducers({
  maxBeerPriceCategory,
  bars,
});

export default beersUnderXApp;

// Local Variables:
// mode: js2-jsx
// End:
