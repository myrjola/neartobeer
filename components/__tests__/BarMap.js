// Import both of these before the test renderer
import React from 'react';
import 'react-native';

import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import mockStore from 'redux-mock-store';
import { initialState as walkingDirectionsInitialState } from '../../reducers/walkingDirections';

import BarMap from '../BarMap';

jest.mock('../../containers/WalkingDirections', () => 'WalkingDirections');

const exampleBars = [
  {
    post_id: 38,
    post_title: 'MacLarens',
    post_author: 1,
    post_content: 'Serves chilled Norrlands Guld on tap for only 29:-.' +
      'One of the most best places in town for cheap beer lovers.',
    post_category: 'Beer under 30',
    default_category: 'Beer under 30',
    post_type: 'gd_place',
    post_status: 'publish',
    post_address: 'Norrtullsgatan 11',
    post_city: 'Stockholm',
    post_region: 'Stockholm County',
    post_country: 'Sweden',
    post_zip: '113 29',
    post_latitude: 59.3419634,
    post_longitude: 18.051719899999966,
  },
  {
    post_id: 39,
    post_title: 'Lion Bar Sveavägen 39',
    post_author: 1,
    post_content: 'The famous Lion Bar chain is a perfect find for bargain brew lovers! 7 Places all around' +
      ' Stockholm serve chilled Spendrups (40cl, tapped) for 23:- until 20h. After this the prices will increase ' +
      'in two steps: until midnight the price will be 36:-, thereafter 40:-. Go and get there early to get yourself' +
      ' a crispy brew for under 30:- !',
    post_category: 'Beer under 50',
    default_category: 'Beer under 30',
    post_type: 'gd_place',
    post_status: 'publish',
    post_address: 'Sveavägen 39 Stockholm',
    post_city: 'Stockholm',
    post_region: 'Stockholm County',
    post_country: 'Sweden',
    post_latitude: 59.3404626,
    post_longitude: 18.05888660000005,
  },
];

var store = null;

beforeEach(() => {
  store = mockStore({ walkingDirections: walkingDirectionsInitialState });
});

it('renders without bars', () => {
  expect(renderer.create(
    <Provider store={store}>
      <BarMap dispatch={jest.fn()} />
    </Provider>,
  )).toMatchSnapshot();
});

// Local Variables:
// mode: js2-jsx
// End:
