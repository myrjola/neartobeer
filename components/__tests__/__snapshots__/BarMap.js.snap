// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with bars 1`] = `
<MapView
  initialRegion={
    Object {
      "latitude": 59.329323,
      "latitudeDelta": 0.0922,
      "longitude": 18.06858,
      "longitudeDelta": 0.0421,
    }
  }
  showsMyLocationButton={false}
  showsUserLocation={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <BarMarkerContainer
    bar={
      Object {
        "default_category": "Beer under 30",
        "post_address": "Norrtullsgatan 11",
        "post_author": 1,
        "post_category": "Beer under 30",
        "post_city": "Stockholm",
        "post_content": "Serves chilled Norrlands Guld on tap for only 29:-.One of the most best places in town for cheap beer lovers.",
        "post_country": "Sweden",
        "post_id": 38,
        "post_latitude": 59.3419634,
        "post_longitude": 18.051719899999966,
        "post_region": "Stockholm County",
        "post_status": "publish",
        "post_title": "MacLarens",
        "post_type": "gd_place",
        "post_zip": "113 29",
      }
    }
  />
  <BarMarkerContainer
    bar={
      Object {
        "default_category": "Beer under 30",
        "post_address": "Sveavägen 39 Stockholm",
        "post_author": 1,
        "post_category": "Beer under 50",
        "post_city": "Stockholm",
        "post_content": "The famous Lion Bar chain is a perfect find for bargain brew lovers! 7 Places all around Stockholm serve chilled Spendrups (40cl, tapped) for 23:- until 20h. After this the prices will increase in two steps: until midnight the price will be 36:-, thereafter 40:-. Go and get there early to get yourself a crispy brew for under 30:- !",
        "post_country": "Sweden",
        "post_id": 39,
        "post_latitude": 59.3404626,
        "post_longitude": 18.05888660000005,
        "post_region": "Stockholm County",
        "post_status": "publish",
        "post_title": "Lion Bar Sveavägen 39",
        "post_type": "gd_place",
      }
    }
  />
  <WalkingDirections />
</MapView>
`;

exports[`renders without bars 1`] = `
<MapView
  initialRegion={
    Object {
      "latitude": 59.329323,
      "latitudeDelta": 0.0922,
      "longitude": 18.06858,
      "longitudeDelta": 0.0421,
    }
  }
  showsMyLocationButton={false}
  showsUserLocation={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <WalkingDirections />
</MapView>
`;
